File: test/stats/statcmp.py
from numpy import *
from scipy import stats
import json, locale
from optparse import OptionParser

VALID_GROUP_BYS = ['browser', 'pdf', 'page', 'round', 'stat']
USAGE_EXAMPLE = "%prog BASELINE CURRENT"
class TestOptions(OptionParser):
    def __init__(self, **kwargs):
        OptionParser.__init__(self, **kwargs)
        self.add_option("--groupBy", action="append", dest="groupBy", type="string",
                        help="How the statistics should grouped. Valid options: " + ', '.join(VALID_GROUP_BYS) + '.', default=[])

        self.set_usage(USAGE_EXAMPLE)

    def verifyOptions(self, options, args):
        if len(args) < 2:
            self.error('There must be two comparison files arguments.')
        # Veryify the group by options.
        groupBy = []
        if not options.groupBy:
          options.groupBy = ['browser,stat', 'browser,pdf,stat']
        for group in options.groupBy:
            group = group.split(',')
            for column in group:
              if column not in VALID_GROUP_BYS:
                self.error('Invalid group by option of "' + column + '"')
            groupBy.append(group)
        options.groupBy = groupBy

        return options

## {{{ http://code.activestate.com/recipes/267662/ (r7)
import cStringIO,operator

def indent(rows, hasHeader=False, headerChar='-', delim=' | ', justify='left',
           separateRows=False, prefix='', postfix='', wrapfunc=lambda x:x):
    """Indents a table by column.
       - rows: A sequence of sequences of items, one sequence per row.
       - hasHeader: True if the first row consists of the columns' names.
       - headerChar: Character to be used for the row separator line
         (if hasHeader==True or separateRows==True).
       - delim: The column delimiter.
       - justify: Determines how are data justified in their column. 
         Valid values are 'left','right' and 'center'.
       - separateRows: True if rows are to be separated by a line
         of 'headerChar's.
       - prefix: A string prepended to each printed row.
       - postfix: A string appended to each printed row.
       - wrapfunc: A function f(text) for wrapping text; each element in
         the table is first wrapped by this function."""
    # closure for breaking logical rows to physical, using wrapfunc
    def rowWrapper(row):
        newRows = [wrapfunc(str(item)).split('\n') for item in row]
        return [[substr or '' for substr in item] for item in map(None,*newRows)]
    # break each logical row into one or more physical ones
    logicalRows = [rowWrapper(row) for row in rows]
    # columns of physical rows
    columns = map(None,*reduce(operator.add,logicalRows))
    # get the maximum of each column by the string length of its items
    maxWidths = [max([len(str(item)) for item in column]) for column in columns]
    rowSeparator = headerChar * (len(prefix) + len(postfix) + sum(maxWidths) + \
                                 len(delim)*(len(maxWidths)-1))
    # select the appropriate justify method
    justify = {'center':str.center, 'right':str.rjust, 'left':str.ljust}[justify.lower()]
    output=cStringIO.StringIO()
    if separateRows: print >> output, rowSeparator
    for physicalRows in logicalRows:
        for row in physicalRows:
            print >> output, \
                prefix \
                + delim.join([justify(str(item),width) for (item,width) in zip(row,maxWidths)]) \
                + postfix
        if separateRows or hasHeader: print >> output, rowSeparator; hasHeader=False
    return output.getvalue()

# written by Mike Brown
# http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/148061
def wrap_onspace(text, width):
    """
    A word-wrap function that preserves existing line breaks
    and most spaces in the text. Expects that existing line
    breaks are posix newlines (\n).
    """
    return reduce(lambda line, word, width=width: '%s%s%s' %
                  (line,
                   ' \n'[(len(line[line.rfind('\n')+1:])
                         + len(word.split('\n',1)[0]
                              ) >= width)],
                   word),
                  text.split(' ')
                 )

import re
def wrap_onspace_strict(text, width):
    """Similar to wrap_onspace, but enforces the width constraint:
       words longer than width are split."""
    wordRegex = re.compile(r'\S{'+str(width)+r',}')
    return wrap_onspace(wordRegex.sub(lambda m: wrap_always(m.group(),width),text),width)

import math
def wrap_always(text, width):
    """A simple word-wrap function that wraps text on exactly width characters.
       It doesn't split the text in words."""
    return '\n'.join([ text[width*i:width*(i+1)] \
                       for i in xrange(int(math.ceil(1.*len(text)/width))) ])

def formatTime(time):
    return locale.format("%.*f", (0, time), True)

# Group the stats by keys. We should really just stick these in a SQL database
# so we aren't reiventing the wheel.
def group(stats, groupBy):
    vals = {}
    for stat in stats:
        key = []
        for group in groupBy:
            key.append(stat[group])
        key = tuple(key)
        if key not in vals:
            vals[key] = []
        vals[key].append(stat['time'])
    return vals;


def mean(l):
    return array(l).mean()


# Take the somewhat normalized stats file and flatten it so there is a row for
# every recorded stat.
def flatten(stats):
    rows = []
    for stat in stats:
        for s in stat['stats']:
            rows.append({
                'browser': stat['browser'],
                'page': stat['page'],
                'pdf': stat['pdf'],
                'round': stat['round'],
                'stat': s['name'],
                'time': int(s['end']) - int(s['start'])
            })
    return rows

# Dump various stats in a table to compare the baseline and current results.
# T-test Refresher:
# If I understand t-test correctly, p is the probability that we'll observe
# another test that is as extreme as the current result assuming the null
# hypothesis is true. P is NOT the probability of the null hypothesis.
# The null hypothesis in this case is that the baseline and current results will
# be the same. It is generally accepted that you can reject the null hypothesis
# if the p-value is less than 0.05.  So if p < 0.05 we can reject the results
# are the same which doesn't necessarily mean the results are faster/slower but
# it can be implied.
def stat(baseline, current, groupBy):
    labels = groupBy + ['Baseline(ms)', 'Current(ms)', '+/-', '%', 'Result(P<.05)']
    baselineGroup = group(baseline, groupBy)
    currentGroup = group(current, groupBy)
    rows = []
    for key in baselineGroup:
        t, p = stats.ttest_ind(baselineGroup[key], currentGroup[key], equal_var = False)
        baseline = mean(baselineGroup[key])
        current = mean(currentGroup[key])
        speed = ''
        if p < 0.05:
          speed = 'faster' if current < baseline else 'slower'
        row = list(key)
        row += [
            formatTime(baseline),
            formatTime(current),
            formatTime(baseline - current),
            round(100 * (1.0 * baseline - current) / baseline, 2),
            speed
        ]
        rows.append(row)
    rows.sort(key=lambda row: tuple(row[0:len(groupBy)]))
    print indent([labels] + rows, hasHeader=True)

def main():
    optionParser = TestOptions()
    options, args = optionParser.parse_args()
    options = optionParser.verifyOptions(options, args)
    if options == None:
        sys.exit(1)

    with open(args[0]) as baselineFile:
        baseline = flatten(json.load(baselineFile))
    with open(args[1]) as currentFile:
        current = flatten(json.load(currentFile))

    for groupBy in options.groupBy:
      print "-- Grouped By " + ', '.join(groupBy) + ' -- '
      stat(baseline, current, groupBy)


if __name__ == '__main__':
    main()

File: test/stats/statcmp.js
/*jslint node: true */

'use strict';

var fs = require('fs');

try {
  var ttest = require('ttest');
} catch (e) {
  console.log('\nttest is not installed -- to intall, run "npm install ttest"');
  console.log('Continuing without significance test...\n');
}

var VALID_GROUP_BYS = ['browser', 'pdf', 'page', 'round', 'stat'];

function parseOptions() {
  var yargs = require('yargs')
    .usage('Compare the results of two stats files.\n' +
           'Usage:\n  $0 <BASELINE> <CURRENT> [options]')
    .demand(2)
    .string(['groupBy'])
    .describe('groupBy', 'How statistics should grouped. Valid options: ' +
              VALID_GROUP_BYS.join(' '))
    .default('groupBy', 'browser,stat');
  var result = yargs.argv;
  result.baseline = result._[0];
  result.current = result._[1];
  if (result.groupBy) {
    result.groupBy = result.groupBy.split(/[;, ]+/);
  }
  return result;
}

function group(stats, groupBy) {
  var vals = [];
  for (var i = 0; i < stats.length; i++) {
    var stat = stats[i];
    var keyArr = [];
    for (var j = 0; j < groupBy.length; j++) {
      keyArr.push(stat[groupBy[j]]);
    }
    var key = keyArr.join(',');
    if (vals[key] === undefined) {
      vals[key] = [];
    }
    vals[key].push(stat['time']);
  }
  return vals;
}

/*
 * Flatten the stats so that there's one row per stats entry.
 * Also, if results are not grouped by 'stat', keep only 'Overall' results.
 */
function flatten(stats) {
  var rows = [];
  stats.forEach(function(stat) {
    stat['stats'].forEach(function(s) {
      rows.push({
        browser: stat['browser'],
        page: stat['page'],
        pdf: stat['pdf'],
        round: stat['round'],
        stat: s['name'],
        time: s['end'] - s['start']
      });
    });
  });
  // Use only overall results if not grouped by 'stat'
  if (options.groupBy.indexOf('stat') < 0) {
    rows = rows.filter(function(s) { return s.stat === 'Overall'; });
  }
  return rows;
}

function pad(s, length, dir /* default: 'right' */) {
  s = '' + s;
  var spaces = new Array(Math.max(0, length - s.length + 1)).join(' ');
  return dir === 'left' ? spaces + s : s + spaces;
}

function mean(array) {
  function add(a, b) {
    return a + b;
  }
  return array.reduce(add, 0) / array.length;
}

/* Comparator for row key sorting. */
function compareRow(a, b) {
  a = a.split(',');
  b = b.split(',');
  for (var i = 0; i < Math.min(a.length, b.length); i++) {
    var intA = parseInt(a[i], 10);
    var intB = parseInt(b[i], 10);
    var ai = isNaN(intA) ? a[i] : intA;
    var bi = isNaN(intB) ? b[i] : intB;
    if (ai < bi) {
      return -1;
    }
    if (ai > bi) {
      return 1;
    }
  }
  return 0;
}

/*
 * Dump various stats in a table to compare the baseline and current results.
 * T-test Refresher:
 * If I understand t-test correctly, p is the probability that we'll observe
 * another test that is as extreme as the current result assuming the null
 * hypothesis is true. P is NOT the probability of the null hypothesis. The null
 * hypothesis in this case is that the baseline and current results will be the
 * same. It is generally accepted that you can reject the null hypothesis if the
 * p-value is less than 0.05. So if p < 0.05 we can reject the results are the
 * same which doesn't necessarily mean the results are faster/slower but it can
 * be implied.
 */
function stat(baseline, current) {
  var baselineGroup = group(baseline, options.groupBy);
  var currentGroup = group(current, options.groupBy);

  var keys = Object.keys(baselineGroup);
  keys.sort(compareRow);

  var labels = options.groupBy.slice(0);
  labels.push('Count', 'Baseline(ms)', 'Current(ms)', '+/-', '% ');
  if (ttest) {
    labels.push('Result(P<.05)');
  }
  var i, row, rows = [];
  // collect rows and measure column widths
  var width = labels.map(function(s) { return s.length; });
  rows.push(labels);
  for (var k = 0; k < keys.length; k++) {
    var key = keys[k];
    var baselineMean = mean(baselineGroup[key]);
    var currentMean = mean(currentGroup[key]);
    row = key.split(',');
    row.push('' + baselineGroup[key].length,
             '' + Math.round(baselineMean),
             '' + Math.round(currentMean),
             '' + Math.round(currentMean - baselineMean),
             (100 * (currentMean - baselineMean) / baselineMean).toFixed(2));
    if (ttest) {
      var p = (baselineGroup[key].length < 2) ? 1 :
               ttest(baselineGroup[key], currentGroup[key]).pValue();
      if (p < 0.05) {
        row.push(currentMean < baselineMean ? 'faster' : 'slower');
      } else {
        row.push('');
      }
    }
    for (i = 0; i < row.length; i++) {
      width[i] = Math.max(width[i], row[i].length);
    }
    rows.push(row);
  }

  // add horizontal line
  var hline = width.map(function(w) { return new Array(w+1).join('-'); });
  rows.splice(1, 0, hline);

  // print output
  console.log('-- Grouped By ' + options.groupBy.join(', ') + ' --');
  var groupCount = options.groupBy.length;
  for (var r = 0; r < rows.length; r++) {
    row = rows[r];
    for (i = 0; i < row.length; i++) {
      row[i] = pad(row[i], width[i], (i < groupCount) ? 'right' : 'left');
    }
    console.log(row.join(' | '));
  }
}

function main() {
  var baseline, current;
  try {
    var baselineFile = fs.readFileSync(options.baseline).toString();
    baseline = flatten(JSON.parse(baselineFile));
  } catch(e) {
    console.log('Error reading file "' + options.baseline + '": ' + e);
    process.exit(0);
  }
  try {
    var currentFile = fs.readFileSync(options.current).toString();
    current = flatten(JSON.parse(currentFile));
  } catch(e) {
    console.log('Error reading file "' + options.current + '": ' + e);
    process.exit(0);
  }
  stat(baseline, current);
}

var options = parseOptions();
main();
